File: frontend/src/api/board-agents.ts
/**
 * Board–Agent membership API (batch add/remove).
 *
 * These endpoints manage the agent_boards M2M relationship without triggering
 * gateway reprovisioning — they are pure metadata operations.
 */

import { customFetch } from "./mutator";

export interface BoardAgentsBatchPayload {
  agent_ids: string[];
}

export interface BoardAgentsBatchResult {
  added: string[];
  removed: string[];
  skipped: string[];
}

interface ApiResponse<T> {
  data: T;
  status: number;
  headers: Headers;
}

/**
 * Batch-add existing agents to a board.
 */
export async function addAgentsToBoard(
  boardId: string,
  payload: BoardAgentsBatchPayload,
): Promise<ApiResponse<BoardAgentsBatchResult>> {
  return customFetch<ApiResponse<BoardAgentsBatchResult>>(
    `/api/v1/boards/${boardId}/agents`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
}

/**
 * Batch-remove agents from a board.
 */
export async function removeAgentsFromBoard(
  boardId: string,
  payload: BoardAgentsBatchPayload,
): Promise<ApiResponse<BoardAgentsBatchResult>> {
  return customFetch<ApiResponse<BoardAgentsBatchResult>>(
    `/api/v1/boards/${boardId}/agents`,
    {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
}

File: backend/app/schemas/board_agents.py
"""Schemas for batch board–agent membership operations."""

from __future__ import annotations

from uuid import UUID

from sqlmodel import SQLModel, Field


class BoardAgentsBatchPayload(SQLModel):
    """Payload for batch add/remove of agents on a board."""

    agent_ids: list[UUID] = Field(
        description="Agent UUIDs to add or remove.",
        min_length=1,
    )


class BoardAgentsBatchResult(SQLModel):
    """Result summary for a batch board–agent operation."""

    added: list[UUID] = Field(default_factory=list, description="Agent IDs successfully added.")
    removed: list[UUID] = Field(default_factory=list, description="Agent IDs successfully removed.")
    skipped: list[UUID] = Field(
        default_factory=list,
        description="Agent IDs skipped (already present for add, or not on board for remove).",
    )
